### PR TITLE
dts: stm32f7: Remove DTCM from sram0

### LINKED ISSUE #15909: stm32f7: DTCM included in sram0
**Describe the bug**
The SRAM base address and sizes for the STM32F7 dts configs includes DTCM memory @ 0x20000000, but TCM is a different beast and probably shouldn't be lumped with in with SRAM.

**Expected behavior**
DTCM and ITCM require some special handling and have some special features. For the time being, use only SRAM1 and SRAM2.

A feature enhancement would be to enable ITCM and DTCM and make use of them.

**Impact**
May create subtle issues with DMA (LTDC, Ethernet, others fight with CPU). There may be issues with memory blocks that cross the DTCM/SRAM address.

**Environment (please complete the following information):**
STM32F7-based boards, at least.

**Additional context**
ST AN4667:

https://www.st.com/content/ccc/resource/technical/document/application_note/0e/53/06/68/ef/2f/4a/cd/DM00169764.pdf/files/DM00169764.pdf/jcr:content/translations/en.DM00169764.pdf

### PR DESCRIPTION
**Pull request by matthew-koch**
Change sram0 start address to SRAM1 from DTCM.

Fixes: #15909 

Signed-off-by: Matthew Koch <koch.matthew@gmail.com>

### CODE REVIEW

File: dts/arm/st/f7/stm32f723.dtsi [hunk 1]
```diff
@@ -7,6 +7,14 @@
 #include <st/f7/stm32f7.dtsi>
 
 / {
+	/* 64KB DTCM @ 0x20000000, 176KB SRAM1 @ 0x20010000, 16KB SRAM2 @ 0x2003C00 */
+
+	sram0: memory@20010000 {
+		device_type = "memory";
+		compatible = "mmio-sram";
+		reg = <0x20010000 DT_SIZE_K(196)>;
```
Comment (matthew-koch) [inline dts/arm/st/f7/stm32f723.dtsi]: I can maths...

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this change, I'm surprised we haven't come across issues due to that before.
In order to avoid inconsistencies and potential error at later stages, please add a commit to update `ram` section in yaml files for impacted boards.

File: dts/arm/st/f7/stm32f769.dtsi [hunk 2]
```diff
@@ -1,7 +1,54 @@
 /*
- * Copyright (c)  2018 Yong Jin
+ * Copyright (c) 2018 Yurii Hamann
```
[the remark this answers is not in the document]
Comment (matthew-koch) [reply dts/arm/st/f7/stm32f769.dtsi]: I nuked Yong's file and replaced it with a slightly modified file from Yurii.

**Comment by matthew-koch** [comment]
> Thanks for this change, I'm surprised we haven't come across issues due to that before.

Not sure there's a lot of use of these chips (yet). It bit me right away working on LTDC.

> In order to avoid inconsistencies and potential error at later stages, please add a commit to update `ram` section in yaml files for impacted boards.

I think I got them all. Looks like there are only ST Nucleo and Discovery boards. I haven't seem the board YAML files used for anything in the build process. Will have to check that out.

**Comment by aurel32** [comment]
> Thanks for this change, I'm surprised we haven't come across issues due to that before.
> In order to avoid inconsistencies and potential error at later stages, please add a commit to update `ram` section in yaml files for impacted boards.

Well so far it actually had a positive impact, for example it was the reason why the STM32F7 family was not impacted by #8786 or #12548 ... At least for the DTCM part of the memory.

**Comment by erwango** [comment]
> Well so far it actually had a positive impact, for example it was the reason why the STM32F7 family was not impacted by #8786 or #12548 ... At least for the DTCM part of the memory.

Great!
From what I see, these have been solved or are taken into account on STM32F7. Any other impact that should be taken into account?

File: dts/arm/st/f7/stm32f769.dtsi (hunk 2 above)
[the remark this answers is not in the document]
Comment (erwango) [reply dts/arm/st/f7/stm32f769.dtsi]: (but one could argue that you made the decision for stm32f769.dtsi file which has not been considered by Yurii...)
Anyway: ok.

**Comment by aurel32** [comment]
> > Well so far it actually had a positive impact, for example it was the reason why the STM32F7 family was not impacted by #8786 or #12548 ... At least for the DTCM part of the memory.
> 
> Great!
> From what I see, these have been solved or are taken into account on STM32F7. Any other impact that should be taken into account?

Yes those issues have been solved and I am not aware of any remaining issues. It's mostly that the DTCM memory hides all cache issues, and I agree it's a bad idea to have a memory range than span DTCM and normal memory. Therefore the change looks good to me in principle (I haven't reviewed the patches details yet). The consequence of removing the DTCM memory is that more MPU memory regions might be needed, we have to pay attention that Zephyr stills build for all the F7 SoCs when the MPU is enabled.

It might also be a good idea to open a wishlist issue to add proper support for DTCM memory, or at least to leave users choose between SRAM and DTCM. This is not something specific to the F7 SoCs and will also benefits the Atmel SAM E70 and NXP iMXRT10xx SoCs.

**Comment by matthew-koch** [comment]
Just want to ping this PR and make sure you aren't waiting on anything from me.

**Review by aurel32 — APPROVED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
Looks fine.
Minor change remains: please update title for the second commit.

**Comment by matthew-koch** [comment]
> Looks fine.
> Minor change remains: please update title for the second commit.

Uh, oh. I have to claim a little ignorance here, but want to learn from this. Should it be, "boards: stm32f7: Remove DTCM from SRAM" or am I missing something else?

**Comment by erwango** [comment]
> Should it be, "boards: stm32f7: Remove DTCM from SRAM" or am I missing something else?

This will be fine

**Review by erwango — APPROVED** [review]

**Comment by galak** [comment]
Should we backport this to 1.14?

**Review by nashif — APPROVED** [review]

**Pull request merged**